File: src/rac/core/identity.py
"""The deterministic artifact identifier (ADR-026).

Identity is a shared core primitive: repository indexing, relationship
resolution, and portfolio analysis all ask this module "what does this artifact
answer to". It is pure and deterministic (ADR-002) — it reads only the parsed
:class:`~rac.core.models.Product` and the file path, never git, the clock, or any
other external state.

One precedence chain drives everything here: frontmatter ``id`` > an explicit
``## ID`` section > the type's declared ``spec.id_field`` > a ``<letters>-<digits>``
prefix of the filename stem > the whole stem. :func:`artifact_identifier` returns
the winner, :func:`artifact_identifiers` returns the whole chain deduped (the
migration aliases), and :func:`identity_conflict` flags a frontmatter id that
disagrees with a declared legacy id.
"""

from __future__ import annotations

import re
from collections.abc import Iterator
from pathlib import Path

from .artifacts import ArtifactSpec
from .models import Product

# A well-formed leading Markdown list marker — ``-``, ``*``, ``+`` or ``N.``
# plus whitespace — stripped from the first line of a single-value section so
# ``- ADR-099`` and ``ADR-099`` yield the same identity.
_LIST_MARKER_RE = re.compile(r"^(?:[-*+]|\d+\.)\s+")

# A recognised leading ID prefix in a filename stem: <letters>-<digits>, e.g.
# ``adr-004`` from ``adr-004-parser-strategy``. Compared case-insensitively.
_ID_PREFIX_RE = re.compile(r"^[A-Za-z]+-\d+")

# The universal explicit-identifier section, as a normalised heading key.
_ID_SECTION = "id"


def _first_value(body: str | None) -> str:
    """The first non-empty line of a section body, one list marker stripped."""
    if not body:
        return ""
    for line in body.splitlines():
        stripped = line.strip()
        if stripped:
            return _LIST_MARKER_RE.sub("", stripped, count=1).strip()
    return ""


def _legacy_identifier(product: Product, spec: ArtifactSpec | None) -> str:
    """The declared legacy identity: ``## ID`` first, then ``spec.id_field``."""
    explicit = _first_value(product.sections.get(_ID_SECTION))
    if explicit:
        return explicit
    if spec is not None and spec.id_field:
        declared = _first_value(product.sections.get(spec.id_field))
        if declared:
            return declared
    return ""


def _candidate_identifiers(product: Product, spec: ArtifactSpec | None, path: str) -> Iterator[str]:
    """Yield every identity candidate in precedence order (may include blanks).

    The single source of truth for the precedence chain: the canonical
    frontmatter id (already uppercase-normalised by the frontmatter parser),
    then the declared legacy id, then the filename-stem prefix (only when one
    matches), then the whole stem. Blanks are yielded as-is; the callers decide
    whether to skip them.
    """
    if product.metadata is not None and product.metadata.id:
        yield product.metadata.id
    yield _legacy_identifier(product, spec)
    stem = Path(path).stem
    prefix = _ID_PREFIX_RE.match(stem)
    if prefix:
        yield prefix.group(0)
    yield stem


def artifact_identifier(product: Product, spec: ArtifactSpec | None, path: str) -> str:
    """The single canonical identifier for the artifact at ``path``.

    The first non-blank candidate in the precedence chain wins (see
    :func:`_candidate_identifiers`); an empty stem with nothing else declared
    falls through to that same empty stem. Conflicts between frontmatter and
    legacy identity are *not* resolved here — :func:`identity_conflict` detects
    them and validation reports them; this function answers only "which identity
    is canonical" (frontmatter wins).
    """
    for candidate in _candidate_identifiers(product, spec, path):
        if candidate:
            return candidate
    return Path(path).stem


def artifact_identifiers(product: Product, spec: ArtifactSpec | None, path: str) -> list[str]:
    """Every identifier this artifact answers to, canonical first, deduped.

    The canonical identifier leads (the value :func:`artifact_identifier`
    returns); the declared legacy id, filename prefix, and whole stem follow as
    migration aliases. Reference resolution indexes all of them so an existing
    human-readable reference (e.g. ``ADR-015``) keeps resolving after an artifact
    adopts a canonical frontmatter id. Duplicates are dropped case-insensitively,
    and duplicate-identity detection elsewhere uses only the canonical (first)
    value, so an alias never manufactures a duplicate on its own.
    """
    ids: list[str] = []
    seen: set[str] = set()
    for candidate in _candidate_identifiers(product, spec, path):
        folded = candidate.casefold()
        if candidate and folded not in seen:
            ids.append(candidate)
            seen.add(folded)
    return ids


def identity_conflict(product: Product, spec: ArtifactSpec | None) -> tuple[str, str] | None:
    """Detect a frontmatter id that disagrees with a declared legacy id.

    Returns ``(frontmatter_id, legacy_id)`` when both are declared and differ
    (compared case-insensitively — matching values are accepted so an artifact
    can carry both during migration), otherwise None. Filename-derived identity
    is a fallback, not a declaration, so it never conflicts.
    """
    if product.metadata is None or not product.metadata.id:
        return None
    legacy = _legacy_identifier(product, spec)
    if not legacy:
        return None
    if legacy.strip().upper() == product.metadata.id:
        return None
    return (product.metadata.id, legacy)

File: rust/rac-engine/src/identity.rs
//! Artifact identity (`rac.core.identity`), per PORT-CONTRACT.d/04 §3.
//!
//! Precedence: frontmatter id -> `## ID` section -> `spec.id_field` (dead
//! today) -> filename-stem `^[A-Za-z]+-\d+` prefix -> whole stem. Pathlib
//! `stem` semantics and Unicode `\d` are replicated exactly.

use crate::parse::Artifact;
use crate::pycompat::{is_re_digit, py_casefold, py_is_space, py_splitlines, py_strip};
use crate::spec::ArtifactSpec;

/// Strip ONE leading well-formed Markdown list marker
/// (`^(?:[-*+]|\d+\.)\s+`, Unicode `\d`/`\s`) from `s`, or return it as-is.
pub fn strip_list_marker(s: &str) -> &str {
    let mut chars = s.char_indices();
    let Some((_, first)) = chars.next() else {
        return s;
    };
    let after_marker = if matches!(first, '-' | '*' | '+') {
        first.len_utf8()
    } else if is_re_digit(first) {
        // \d+ then a literal '.'
        let mut end = first.len_utf8();
        let mut rest = s[end..].char_indices();
        loop {
            match rest.next() {
                Some((_, c)) if is_re_digit(c) => end += c.len_utf8(),
                Some((_, '.')) => {
                    end += 1;
                    break;
                }
                _ => return s,
            }
        }
        end
    } else {
        return s;
    };
    // \s+ — at least one Python-whitespace char.
    let tail = &s[after_marker..];
    let trimmed = tail.trim_start_matches(py_is_space);
    if trimmed.len() == tail.len() {
        return s; // no whitespace after the marker -> no match
    }
    trimmed
}

/// identity's `_first_value(body)`: first non-empty stripped line, one
/// leading list marker stripped, stripped again.
pub fn first_value_list_stripped(body: Option<&str>) -> String {
    let Some(body) = body else {
        return String::new();
    };
    if body.is_empty() {
        return String::new();
    }
    for line in py_splitlines(body) {
        let stripped = py_strip(line);
        if !stripped.is_empty() {
            return py_strip(strip_list_marker(stripped)).to_string();
        }
    }
    String::new()
}

/// `Path(path).stem` — pathlib semantics: final component (trailing slashes
/// and `.` segments dropped), minus the last suffix, where a suffix exists
/// only when the final dot is neither at index 0 nor the last char.
pub fn path_stem(path: &str) -> String {
    let name = path
        .split('/')
        .rfind(|p| !p.is_empty() && *p != ".")
        .unwrap_or("");
    match name.rfind('.') {
        Some(i) if i > 0 && i < name.len() - 1 => name[..i].to_string(),
        _ => name.to_string(),
    }
}

/// `_ID_PREFIX_RE = ^[A-Za-z]+-\d+` (ASCII letters, Unicode digits) —
/// returns the matched prefix (`.group(0)`) or None.
pub fn id_prefix(stem: &str) -> Option<&str> {
    let bytes = stem.as_bytes();
    let mut i = 0;
    while i < bytes.len() && bytes[i].is_ascii_alphabetic() {
        i += 1;
    }
    if i == 0 || i >= bytes.len() || bytes[i] != b'-' {
        return None;
    }
    // \d+ — a contiguous, greedy digit run from digits_start.
    let digits_start = i + 1;
    let mut run_end = digits_start;
    for c in stem[digits_start..].chars() {
        if is_re_digit(c) {
            run_end += c.len_utf8();
        } else {
            break;
        }
    }
    if run_end == digits_start {
        None
    } else {
        Some(&stem[..run_end])
    }
}

/// `_legacy_identifier(product, spec)`: `## ID` first value, then
/// `spec.id_field` (no spec sets it — ported for fidelity).
fn legacy_identifier(artifact: &Artifact, spec: Option<&ArtifactSpec>) -> String {
    let explicit = first_value_list_stripped(artifact.section("id"));
    if !explicit.is_empty() {
        return explicit;
    }
    if let Some(spec) = spec {
        if let Some(field) = &spec.id_field {
            if !field.is_empty() {
                let declared = first_value_list_stripped(artifact.section(field));
                if !declared.is_empty() {
                    return declared;
                }
            }
        }
    }
    String::new()
}

fn metadata_id(artifact: &Artifact) -> Option<&str> {
    artifact
        .metadata
        .as_ref()
        .and_then(|m| m.id.as_deref())
        .filter(|s| !s.is_empty())
}

/// `artifact_identifier(product, spec, path)`.
pub fn artifact_identifier(artifact: &Artifact, spec: Option<&ArtifactSpec>, path: &str) -> String {
    if let Some(id) = metadata_id(artifact) {
        return id.to_string();
    }
    let legacy = legacy_identifier(artifact, spec);
    if !legacy.is_empty() {
        return legacy;
    }
    let stem = path_stem(path);
    match id_prefix(&stem) {
        Some(prefix) => prefix.to_string(),
        None => stem,
    }
}

/// `artifact_identifiers(product, spec, path)`: canonical first, then legacy
/// aliases, de-duplicated case-insensitively (casefold).
pub fn artifact_identifiers(
    artifact: &Artifact,
    spec: Option<&ArtifactSpec>,
    path: &str,
) -> Vec<String> {
    let mut ids: Vec<String> = Vec::new();
    let mut folded: Vec<String> = Vec::new();
    let add = |value: String, ids: &mut Vec<String>, folded: &mut Vec<String>| {
        if value.is_empty() {
            return;
        }
        let f = py_casefold(&value);
        if folded.contains(&f) {
            return;
        }
        folded.push(f);
        ids.push(value);
    };
    if let Some(id) = metadata_id(artifact) {
        add(id.to_string(), &mut ids, &mut folded);
    }
    add(legacy_identifier(artifact, spec), &mut ids, &mut folded);
    let stem = path_stem(path);
    if let Some(prefix) = id_prefix(&stem) {
        add(prefix.to_string(), &mut ids, &mut folded);
    }
    add(stem, &mut ids, &mut folded);
    ids
}

/// `identity_conflict(product, spec)` -> `(frontmatter_id, legacy_id)`.
pub fn identity_conflict(
    artifact: &Artifact,
    spec: Option<&ArtifactSpec>,
) -> Option<(String, String)> {
    let fm_id = metadata_id(artifact)?;
    let legacy = legacy_identifier(artifact, spec);
    if legacy.is_empty() {
        return None;
    }
    // Compared as `legacy.strip().upper() == metadata.id` (upper, not
    // casefold — the asymmetry is deliberate, PORT-CONTRACT.d/04 §3.4).
    if py_strip(&legacy).to_uppercase() == fm_id {
        return None;
    }
    Some((fm_id.to_string(), legacy))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn stem_semantics() {
        assert_eq!(path_stem("a.b.md"), "a.b");
        assert_eq!(path_stem("a."), "a.");
        assert_eq!(path_stem(".hidden"), ".hidden");
        assert_eq!(path_stem("a/b/"), "b");
        assert_eq!(path_stem("a..md"), "a.");
        assert_eq!(path_stem(".md"), ".md");
        assert_eq!(path_stem("tests/fixtures/valid/feature.md"), "feature");
    }

    #[test]
    fn prefix_matching() {
        assert_eq!(id_prefix("adr-004-parser-strategy"), Some("adr-004"));
        assert_eq!(id_prefix("adr-x"), None);
        assert_eq!(id_prefix("-004"), None);
        assert_eq!(id_prefix("adr004"), None);
    }

    #[test]
    fn list_marker_strip() {
        assert_eq!(strip_list_marker("- ADR-1"), "ADR-1");
        assert_eq!(strip_list_marker("12. x"), "x");
        assert_eq!(strip_list_marker("-x"), "-x");
        assert_eq!(strip_list_marker("+  y"), "y");
    }
}
